galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,478 @@
+    (+f2) is at a position angle pa_fpa = pa_obsy + 90 + theta_fpa.
+
+    @param PA        Position angle of the observatory Y axis, unless `PA_is_FPA=True`, in which
+                     case it's the position angle of the FPA.  Must be provided as a galsim.Angle.
+    @param world_pos A galsim.CelestialCoord indicating the center of the FPA.
+    @param PA_is_FPA If True, then the position angle that was provided was the PA of the focal
+                     plane array, not the observatory. [default: False]
+    @param SCAs      A single number or iterable giving the SCAs for which the WCS should be
+                     obtained.  If None, then the WCS is calculated for all SCAs.
+                     [default: None]
+    @returns a dict of WCS objects for each SCA.
+    """
+    # Check which SCAs are to be done using a helper routine in this module.
+    SCAs = galsim.wfirst._parse_SCAs(SCAs)
+
+    # Enforce type for PA
+    if not isinstance(PA, galsim.Angle):
+        raise TypeError("Position angle must be a galsim.Angle!")
+
+    # Parse input position
+    if not isinstance(world_pos, galsim.CelestialCoord):
+        raise TypeError("Position on the sky must be given as a galsim.CelestialCoord!")
+
+    # Compute position angle of FPA f2 axis, where positive corresponds to the angle east of North.
+    if PA_is_FPA:
+        pa_fpa = PA
+        pa_obsy = PA - 90.*galsim.degrees - theta_fpa
+    else:
+        pa_obsy = PA
+        pa_fpa = PA + 90.*galsim.degrees + theta_fpa
+    cos_pa = np.cos(pa_fpa.rad())
+    sin_pa = np.sin(pa_fpa.rad())
+
+    # Figure out tangent-plane positions for FPA center:
+    xc_fpa_tp, yc_fpa_tp = _det_to_tangplane_positions(xc_fpa, yc_fpa)
+
+    # Note, this routine reads in the coeffs.  We don't use them until later, but read them in for
+    # all SCAs at once.
+    a_sip, b_sip = _parse_sip_file(sip_filename)
+
+    # Loop over SCAs:
+    wcs_dict = {}
+    for i_sca in SCAs:
+        # Set up the header.
+        header = galsim.FitsHeader()
+        # Populate some necessary variables in the FITS header that are always the same, regardless of
+        # input and SCA number.
+        _populate_required_fields(header)
+
+        # And populate some things that just depend on the overall locations or other input, not on
+        # the SCA.
+        header['RA_TARG'] = (world_pos.ra / galsim.degrees,
+                             "right ascension of the target (deg) (J2000)")
+        header['DEC_TARG'] = (world_pos.dec / galsim.degrees,
+                              "declination of the target (deg) (J2000)")
+        header['PA_OBSY'] = (pa_obsy / galsim.degrees, "position angle of observatory Y axis (deg)")
+        header['PA_FPA'] = (pa_fpa / galsim.degrees, "position angle of FPA Y axis (deg)")
+
+        # Finally do all the SCA-specific stuff.
+        header['SCA_NUM'] = (i_sca, "SCA number (1 - 18)")
```

Comment:
>  Have you thought about refactoring this to have a separate function that adds appropriate header items to the 
> FITS header?

Yes, I have - it's still on my list of things to do from your original set of suggestions.   I just didn't have time to get to it quite yet, partly due to lack of time (had to go get some stuff done for other projects for a few days) and partly because I tend to leave things for last if they seem kind of annoying to deal with (I hate dealing with FITS header stuff, even for a good cause).

What is your thought on how this routine would be used?
i.e., right now people just save an image to file using ```image.write()``` and the necessary WCS keys get saved in the fits header due to the fact that the ```write()``` method calls  ```image.wcs.writeToFitsHeader()```.  The user never directly sets up a header object.  So are you suggesting that people who want that extra info in the fits header should mess around with making a header and calling this new routine that populates it with those keywords?  Or do you think I should set up some helper routine called something like ```galsim.wfirst.write()``` that acts like the normal ```write()``` method but also calls a helper routine ```galsim.wfirst.populateFitsHeader()``` along the way to add the extra stuff to the header?  In that case the user could then make their image and call ```galsim.wfirst.write(image)``` to save it to file, instead of ```image.write()```.